pkg/scheduler/queue/multi_queuing_algorithm_tree_queue_test.go

```diff
@@ -0,0 +1,982 @@
+		tenantNodes:      make(map[string][]*Node),
+	}
+}
+
+func Test_NewTree(t *testing.T) {
+	tests := []struct {
+		name      string
+		treeAlgos []QueuingAlgorithm
+		state     *tenantQuerierAssignments
+		expectErr bool
+	}{
+		{
+			name:      "create round-robin tree",
+			treeAlgos: []QueuingAlgorithm{&roundRobinState{}},
+		},
+		{
+			name:      "create tenant-querier tree",
+			treeAlgos: []QueuingAlgorithm{&tenantQuerierAssignments{}},
+		},
+		{
+			name:      "fail to create tree without defined dequeuing algorithm",
+			treeAlgos: []QueuingAlgorithm{nil},
+			expectErr: true,
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			_, err := NewTree(tt.treeAlgos...)
+			if tt.expectErr {
+				require.Error(t, err)
+			} else {
+				require.NoError(t, err)
+			}
+
+		})
+	}
+}
+
+func Test_EnqueueBackByPath(t *testing.T) {
+	tests := []struct {
+		name                string
+		treeAlgosByDepth    []QueuingAlgorithm
+		childPathsToEnqueue []QueuePath
+		expectErr           bool
+	}{
+		{
+			name:                "enqueue tenant-querier node to round-robin node",
+			treeAlgosByDepth:    []QueuingAlgorithm{&roundRobinState{}, &roundRobinState{}},
+			childPathsToEnqueue: []QueuePath{{"child-1"}},
+		},
+		{
+			name:                "enqueue tenant-querier node to round-robin node",
+			treeAlgosByDepth:    []QueuingAlgorithm{&roundRobinState{}, &tenantQuerierAssignments{tenantQuerierIDs: map[TenantID]map[QuerierID]struct{}{}}},
+			childPathsToEnqueue: []QueuePath{{"child-1"}, {"child-2"}},
+		},
+		{
+			name: "enqueue round-robin node to tenant-querier node",
+			treeAlgosByDepth: []QueuingAlgorithm{
+				newTenantQuerierAssignments(),
+				&roundRobinState{},
+			},
+			childPathsToEnqueue: []QueuePath{{"child-1"}, {"child-2"}},
+		},
+		{
+			name: "enqueue tenant-querier node to tenant-querier node",
+			treeAlgosByDepth: []QueuingAlgorithm{
+				newTenantQuerierAssignments(),
+				newTenantQuerierAssignments(),
+			},
+			childPathsToEnqueue: []QueuePath{{"child-1"}},
+		},
+		{
+			name: "enqueue grandchildren to a tree with max-depth of 2",
+			treeAlgosByDepth: []QueuingAlgorithm{
+				newTenantQuerierAssignments(),
+				&roundRobinState{},
+				newTenantQuerierAssignments(),
+			},
+			childPathsToEnqueue: []QueuePath{{"child"}, {"grandchild"}},
+		},
+		{
+			name:                "enqueue beyond max-depth",
+			treeAlgosByDepth:    []QueuingAlgorithm{&roundRobinState{}},
+			childPathsToEnqueue: []QueuePath{{"child"}, {"child, grandchild"}},
+			expectErr:           true,
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			tree, err := NewTree(tt.treeAlgosByDepth...)
+			require.NoError(t, err)
+
+			for _, childPath := range tt.childPathsToEnqueue {
+				err = tree.EnqueueBackByPath(childPath, "some-object")
+			}
+			if tt.expectErr {
+				require.Error(t, err)
+			} else {
+				require.NoError(t, err)
+			}
+		})
+	}
+}
+
+func Test_EnqueueFrontByPath(t *testing.T) {
```

Comment:
let's add a short note here on why/what we are testing this particular pattern here - we want to put something that was dequeued but failed to be sent to a consumer to be put back in the front of the line

Reply:
Doesn't it make more sense to document this at the function in question? There's a pretty extensive docstring here: https://github.com/grafana/mimir/blob/44a393daad2547bb497778a6e18ffbb265b1d03c/pkg/scheduler/queue/multi_queuing_algorithm_tree_queue.go#L91-L105

Reply:
yeah that's fine by